twine/utils.py

```diff
@@ -238,29 +238,47 @@ def get_password_from_keyring(system, username):
         warnings.warn(str(exc))
 
 
-def username_from_keyring_or_prompt(system):
+def non_interactive_check_prompt(non_interactive, prompt_func, prompt_type):
+    if non_interactive:
+        error_message = "Credential not found for {}.".format(prompt_type)
+        raise exceptions.NonInteractive(error_message)
+    else:
+        message = "Enter your {}: ".format(prompt_type)
+        return prompt_func(message)
+
+
+def username_from_keyring_or_prompt(system, non_interactive):
     return (
         get_username_from_keyring(system)
-        or input_func('Enter your username: ')
+        or non_interactive_check_prompt(
```

Comment:
The name of this function sounds to me as though it should _always_ be non-interactive. What if instead (since we're cascading this value through so many places) we instead do something like:

```
prompt_func = functools.partial(input_func, "Enter your username: ")
if non_interactive:
    prompt_func = function_that_raises_noninteractive_exc
return (
    get_username_from_keyring(system)
    or prompt_func
)
```

And given that we have similar logic in `password_from_keyring_or_prompt` maybe it makes sense to pass in `prompt_func` instead of passing in `non_interactive` where we can generate the function from something like `generate_prompt_func_from(non_interactive, prompt_msg)`

Reply:
Makes sense, thanks! I believe I've reworked it to align with the last paragraph.